src/analysis/souffle/analyze_tag_checks_test_helper.dl

```diff
@@ -0,0 +1,74 @@
+#ifndef SRC_ANALYSIS_SOUFFLE_ANALYZE_TAG_CHECKS_TEST_HELPER_DL_
+#define SRC_ANALYSIS_SOUFFLE_ANALYZE_TAG_CHECKS_TEST_HELPER_DL_
+
+#include "taint.dl"
+#include "analyze_tag_checks.dl"
+
+// This is a tag check that does not correspond to the original test's arcs manifest. We want to
+// eventually translate the arcs manifests into facts and check them against our hand-written facts,
+// so distinguishing the extra tags will be important eventually. For now, thought, it just works
+// exactly like checkTag.
+.decl checkTagExtra(accessPath: AccessPath, tag: Tag, tagPresence: TagPresence)
+
+checkTag(accessPath, tag, tagPresence) :- checkTagExtra(accessPath, tag, tagPresence).
+
+.type TestResult <: symbol
+.decl isValidTestResult(testResult: TestResult)
+
+isValidTestResult("PASS").
+isValidTestResult("FAIL").
+
+.decl tagCheckExpectation(
+  accessPath: AccessPath,
+  tag: Tag,
+  testResult: TestResult)
+
+.type BadCheckDiagnostic <: symbol
+
+// Used to report malformed test cases.
+.decl badTagCheckExpectation(
```

Comment:
Is this similar to what I call `testFails`, but with more diagnostic pointers like the `accessPath` that can be used to report better test failure messages?

Reply:
It's a bit different from that. These `badTagCheckExpectation` relations are not for capturing whether the test succeeded or failed, but whether the test was valid in the first place.

For instance, if you have stated in a facts file that you expect a particular check to both pass and fail, there is a problem with the test. It almost doesn't make sense to describe it as passing or failing; it isn't coherent. That is what the `badTagCheckExpectation` is trying to find. Currently, it finds the following cases:

1. The test has expectations on a particular check that conflict with each other (expecting it to both pass and fail).
2. The test expresses an expectation that a particular check shall pass or fail, but that check was not one expressed in the test.
3. There is a check in the test that did not have an associated expectation. (This one assumes that, because this is a test, every check is important to the result of the test.)

Reply:
This is now gone, simplified as per your recommendation.